ocaml/xapi/xapi_vdi.ml

```diff
@@ -111,6 +111,17 @@ let check_operation_error ~__context ?(sr_records=[]) ?(pbd_records=[]) ?(vbd_re
             vbd_record.Db_actions.vBD_VDI = _ref' && (vbd_record.Db_actions.vBD_currently_attached || vbd_record.Db_actions.vBD_reserved)
           ) vbd_records)
       in
+      (* Only consider VDIs attached to a non-snapshot VM as actually attached,
+         because VM.checkpoint may set the currently_attached fields of the
+         snapshot's VBDs to true, and this would block operations that require
+         the VDI to be detached. *)
+      let my_active_vbd_records =
+        my_active_vbd_records |> List.filter
+          (fun vbd ->
+             let vm = vbd.Db_actions.vBD_VM in
+             Db.is_valid_ref __context vm && (Db.VM.get_is_a_snapshot ~__context ~self:vm = false)
```

Comment:
Surely `&& not (expression) would be nicer than `&& (expression = false`?

In principle the `vm` could become an invalid ref between the `is_valid_ref` and the `get_is_a_snapshot` calls.
Probably not worth worrying about, but we could handle that situation with something like
```
try
  not (Db.VM.get_is_a_snapshot...)
with InvalidRef -> false
```

Reply:
Oh, but I see your second commit deals with that sort of thing.

Reply:
Oh makes sense, I'll use `not`, thanks.